TestCliGet integration test is flaky

On repeated minikube runs, I've seen `TestCliGet` fail on occasion. It looks like we're inspecting the output of the get command before all pods have initialized.

```
Running test [get_test.go]
=== RUN   TestCliGet
--- FAIL: TestCliGet (0.91s)
	get_test.go:91: Expected conduit get to return:
		[cli-get-test-d1 cli-get-test-d1 cli-get-test-d2 cli-get-test-not-injected-d1 cli-get-test-not-injected-d1 cli-get-test-not-injected-d2]
		But got:
		[cli-get-test-d1 cli-get-test-d1 cli-get-test-d2 cli-get-test-not-injected-d1 cli-get-test-not-injected-d1]
		Raw output:
		kltest-smoke-test/smoke-test-gateway-644dc67498-w9qqv
		kltest-get-test/cli-get-test-not-injected-d1-75595bff74-p9gxf
		kltest-smoke-test/smoke-test-terminus-7778874d5-brdjt
		kltest-get-test/cli-get-test-d2-6f57b58675-5spzj
		kltest-get-test/cli-get-test-not-injected-d1-75595bff74-89cfn
		kltest/grafana-7f4cf4778c-v9rl8
		kltest-get-test/cli-get-test-d1-7d5d6bfc9f-p4dwb
		kltest-egress-test/egress-test-https-post-79c8f9db7f-b5qg6
		kltest-egress-test/egress-test-http-post-76946574c9-x9grq
		kltest/controller-54d9b88bb7-jlrw6
		kltest/web-5d96dbb677-l4v5w
		kltest-egress-test/egress-test-https-get-66f44796d7-w6jrz
		kltest/prometheus-5bfd75647b-tstml
		kltest-get-test/cli-get-test-d1-7d5d6bfc9f-m4bd9
		kltest-egress-test/egress-test-http-get-c949498b7-zd6q9
		kltest-egress-test/egress-test-http-get-75f5bdd57d-qxtgt
FAIL
FAIL	command-line-arguments	1.038s
```

--- a/test/egress/egress_test.go
+++ b/test/egress/egress_test.go
@@ -42,9 +42,9 @@ func TestEgressHttp(t *testing.T) {
 		t.Run(testName, func(t *testing.T) {
 			expectedURL := fmt.Sprintf("%s://%s/%s", protocolToUse, dnsName, strings.ToLower(methodToUse))
 
-			svcURL, err := TestHelper.GetURLForService(prefixedNs, serviceName)
+			svcURL, err := TestHelper.ProxyURLFor(prefixedNs, serviceName, "http")
 			if err != nil {
-				t.Fatalf("Failed to get service URL: %v", err)
+				t.Fatalf("Failed to get proxy URL: %s", err)
 			}
 
 			output, err := TestHelper.HTTPGetURL(svcURL)

--- a/test/egress/testdata/proxy.yaml
+++ b/test/egress/testdata/proxy.yaml
@@ -33,7 +33,6 @@ kind: Service
 metadata:
   name: egress-test-https-post-svc
 spec:
-  type: LoadBalancer
   selector:
     app: egress-test-https-post
   ports:
@@ -74,7 +73,6 @@ kind: Service
 metadata:
   name: egress-test-http-post-svc
 spec:
-  type: LoadBalancer
   selector:
     app: egress-test-http-post
   ports:
@@ -116,7 +114,6 @@ kind: Service
 metadata:
   name: egress-test-https-get-svc
 spec:
-  type: LoadBalancer
   selector:
     app: egress-test-https-get
   ports:
@@ -157,7 +154,6 @@ kind: Service
 metadata:
   name: egress-test-http-get-svc
 spec:
-  type: LoadBalancer
   selector:
     app: egress-test-http-get
   ports:
@@ -199,7 +195,6 @@ kind: Service
 metadata:
   name: egress-test-not-www-get-svc
 spec:
-  type: LoadBalancer
   selector:
     app: egress-test-not-www-get
   ports:

--- a/test/get/get_test.go
+++ b/test/get/get_test.go
@@ -9,6 +9,7 @@ import (
 	"sort"
 	"strings"
 	"testing"
+	"time"
 
 	"github.com/runconduit/conduit/testutil"
 )
@@ -23,6 +24,15 @@ func TestMain(m *testing.M) {
 	TestHelper = testutil.NewTestHelper()
 	os.Exit(m.Run())
 }
+
+var (
+	deployReplicas = map[string]int{
+		"cli-get-test-d1":              2,
+		"cli-get-test-d2":              1,
+		"cli-get-test-not-injected-d1": 2,
+		"cli-get-test-not-injected-d2": 1,
+	}
+)
 
 //////////////////////
 /// TEST EXECUTION ///
@@ -50,6 +60,19 @@ func TestCliGet(t *testing.T) {
 		t.Fatalf("Unexpected error: %v output:\n%s", err, out)
 	}
 
+	// wait for pods to start
+	err = TestHelper.RetryFor(30*time.Second, func() error {
+		for deploy, replicas := range deployReplicas {
+			if err := TestHelper.CheckPods(prefixedNs, deploy, replicas); err != nil {
+				return fmt.Errorf("Error validating pods for deploy [%s]:\n%s", deploy, err)
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		t.Error(err)
+	}
+
 	out, err = TestHelper.ConduitRun("get", "pods")
 	if err != nil {
 		t.Fatalf("Unexpected error: %v output:\n%s", err, out)
@@ -60,8 +83,11 @@ func TestCliGet(t *testing.T) {
 		t.Fatal("Expecting conduit get pods to return something, got nothing")
 	}
 
-	expectedPods := []string{
-		"cli-get-test-d1", "cli-get-test-not-injected-d1", "cli-get-test-not-injected-d1", "cli-get-test-not-injected-d2", "cli-get-test-d1", "cli-get-test-d2",
+	expectedPods := []string{}
+	for deploy, replicas := range deployReplicas {
+		for i := 0; i < replicas; i++ {
+			expectedPods = append(expectedPods, deploy)
+		}
 	}
 
 	var actualPods []string

--- a/test/install_test.go
+++ b/test/install_test.go
@@ -31,7 +31,7 @@ var (
 		"web",
 	}
 
-	conduitDeploys = map[string]int{
+	conduitDeployReplicas = map[string]int{
 		"controller": 1,
 		"grafana":    1,
 		"prometheus": 1,
@@ -83,24 +83,15 @@ func TestInstall(t *testing.T) {
 		t.Error(err)
 	}
 
-	// Tests Pods
-	err = TestHelper.RetryFor(1*time.Minute, func() error {
-		for deploy, replicas := range conduitDeploys {
+	// Tests Pods and Deployments
+	// The Grafana readiness check can take 1+ min; hence the high timeout here
+	err = TestHelper.RetryFor(2*time.Minute, func() error {
+		for deploy, replicas := range conduitDeployReplicas {
 			if err := TestHelper.CheckPods(TestHelper.GetConduitNamespace(), deploy, replicas); err != nil {
 				return fmt.Errorf("Error validating pods for deploy [%s]:\n%s", deploy, err)
 			}
-		}
-		return nil
-	})
-	if err != nil {
-		t.Error(err)
-	}
-
-	// Tests Deployments
-	err = TestHelper.RetryFor(1*time.Minute, func() error {
-		for deploy, replicas := range conduitDeploys {
 			if err := TestHelper.CheckDeployment(TestHelper.GetConduitNamespace(), deploy, replicas); err != nil {
-				return fmt.Errorf("Error validating Deployment [%s]:\n%s", deploy, err)
+				return fmt.Errorf("Error validating deploy [%s]:\n%s", deploy, err)
 			}
 		}
 		return nil

--- a/test/install_test.go
+++ b/test/install_test.go
@@ -22,6 +22,23 @@ func TestMain(m *testing.M) {
 	os.Exit(m.Run())
 }
 
+var (
+	conduitSvcs = []string{
+		"api",
+		"grafana",
+		"prometheus",
+		"proxy-api",
+		"web",
+	}
+
+	conduitDeploys = map[string]int{
+		"controller": 1,
+		"grafana":    1,
+		"prometheus": 1,
+		"web":        1,
+	}
+)
+
 //////////////////////
 /// TEST EXECUTION ///
 //////////////////////
@@ -55,7 +72,7 @@ func TestInstall(t *testing.T) {
 
 	// Tests Services
 	err = TestHelper.RetryFor(10*time.Second, func() error {
-		for _, svc := range []string{"api", "proxy-api", "web"} {
+		for _, svc := range conduitSvcs {
 			if err := TestHelper.CheckService(TestHelper.GetConduitNamespace(), svc); err != nil {
 				return fmt.Errorf("Error validating service [%s]:\n%s", svc, err)
 			}
@@ -67,9 +84,9 @@ func TestInstall(t *testing.T) {
 	}
 
 	// Tests Pods
-	err = TestHelper.RetryFor(30*time.Second, func() error {
-		for _, deploy := range []string{"prometheus", "controller", "web"} {
-			if err := TestHelper.CheckPods(TestHelper.GetConduitNamespace(), deploy, 1); err != nil {
+	err = TestHelper.RetryFor(1*time.Minute, func() error {
+		for deploy, replicas := range conduitDeploys {
+			if err := TestHelper.CheckPods(TestHelper.GetConduitNamespace(), deploy, replicas); err != nil {
 				return fmt.Errorf("Error validating pods for deploy [%s]:\n%s", deploy, err)
 			}
 		}
@@ -80,8 +97,8 @@ func TestInstall(t *testing.T) {
 	}
 
 	// Tests Deployments
-	err = TestHelper.RetryFor(30*time.Second, func() error {
-		for deploy, replicas := range map[string]int{"controller": 1, "prometheus": 1, "web": 1} {
+	err = TestHelper.RetryFor(1*time.Minute, func() error {
+		for deploy, replicas := range conduitDeploys {
 			if err := TestHelper.CheckDeployment(TestHelper.GetConduitNamespace(), deploy, replicas); err != nil {
 				return fmt.Errorf("Error validating Deployment [%s]:\n%s", deploy, err)
 			}
@@ -166,9 +183,9 @@ func TestInject(t *testing.T) {
 		t.Fatalf("kubectl apply command failed\n%s", out)
 	}
 
-	svcURL, err := TestHelper.GetURLForService(prefixedNs, "smoke-test-gateway-svc")
+	svcURL, err := TestHelper.ProxyURLFor(prefixedNs, "smoke-test-gateway-svc", "http")
 	if err != nil {
-		t.Fatalf("Failed to get service URL: %v", err)
+		t.Fatalf("Failed to get proxy URL: %s", err)
 	}
 
 	output, err := TestHelper.HTTPGetURL(svcURL)

--- a/test/testdata/smoke_test.yaml
+++ b/test/testdata/smoke_test.yaml
@@ -60,7 +60,6 @@ metadata:
 spec:
   selector:
     app: smoke-test-gateway
-  type: LoadBalancer
   ports:
   - name: http
     port: 8080

--- a/testutil/kubernetes_helper.go
+++ b/testutil/kubernetes_helper.go
@@ -6,6 +6,7 @@ import (
 	"regexp"
 	"strings"
 
+	"github.com/runconduit/conduit/pkg/k8s"
 	coreV1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
@@ -190,4 +191,23 @@ func (h *KubernetesHelper) ParseNamespacedResource(resource string) (string, str
 		return "", "", fmt.Errorf("string [%s] didn't contain expected format for namespace/resource, extracted: %v", resource, matches)
 	}
 	return matches[0][1], matches[0][2], nil
+}
+
+// ProxyURLFor creates a kubernetes proxy, runs it, and returns the URL that
+// tests can use for access to the given service. Note that the proxy remains
+// running for the duration of the test.
+func (h *KubernetesHelper) ProxyURLFor(namespace, service, port string) (string, error) {
+	proxy, err := k8s.NewProxy("", 0)
+	if err != nil {
+		return "", err
+	}
+
+	url, err := proxy.URLFor(namespace, fmt.Sprintf("/services/%s:%s/proxy/", service, port))
+	if err != nil {
+		return "", err
+	}
+
+	go proxy.Run()
+
+	return url.String(), nil
 }

--- a/testutil/test_helper.go
+++ b/testutil/test_helper.go
@@ -1,7 +1,6 @@
 package testutil
 
 import (
-	"errors"
 	"flag"
 	"fmt"
 	"io/ioutil"
@@ -171,29 +170,6 @@ func (h *TestHelper) CheckVersion(serverVersion string) error {
 	return nil
 }
 
-// BlockUntilTrue retries a given function every second until the function
-// returns true or a timeout is reached. If the timeout is reached, it returns
-// an error.
-func (h *TestHelper) BlockUntilTrue(timeout time.Duration, fn func() bool) error {
-	if fn() {
-		return nil
-	}
-
-	timeoutAfter := time.After(timeout)
-	retryAfter := time.Tick(time.Second)
-
-	for {
-		select {
-		case <-timeoutAfter:
-			return errors.New("timed out waiting for condition")
-		case <-retryAfter:
-			if fn() {
-				return nil
-			}
-		}
-	}
-}
-
 // RetryFor retries a given function every second until the function returns
 // without an error, or a timeout is reached. If the timeout is reached, it
 // returns the last error received from the function.
@@ -221,11 +197,11 @@ func (h *TestHelper) RetryFor(timeout time.Duration, fn func() error) error {
 
 // HTTPGetURL sends a GET request to the given URL. It returns the response body
 // in the event of a successful 200 response. In the event of a non-200
-// response, it returns an error. It retries requests for up to 10 seconds,
+// response, it returns an error. It retries requests for up to 30 seconds,
 // giving pods time to start.
 func (h *TestHelper) HTTPGetURL(url string) (string, error) {
 	var body string
-	err := h.RetryFor(10*time.Second, func() error {
+	err := h.RetryFor(30*time.Second, func() error {
 		resp, err := h.httpClient.Get(url)
 		if err != nil {
 			return err

--- a/testutil/test_helper.go
+++ b/testutil/test_helper.go
@@ -221,58 +221,29 @@ func (h *TestHelper) RetryFor(timeout time.Duration, fn func() error) error {
 
 // HTTPGetURL sends a GET request to the given URL. It returns the response body
 // in the event of a successful 200 response. In the event of a non-200
-// response, it returns an error.
+// response, it returns an error. It retries requests for up to 10 seconds,
+// giving pods time to start.
 func (h *TestHelper) HTTPGetURL(url string) (string, error) {
-	resp, err := h.httpClient.Get(url)
-	if err != nil {
-		// retry once on timeout error; workaround for GKE loadbalancers
-		if strings.Contains(err.Error(), "Client.Timeout") {
-			resp, err = h.httpClient.Get(url)
-		}
-	}
-	if err != nil {
-		return "", err
-	}
-
-	defer resp.Body.Close()
-	bytes, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		return "", fmt.Errorf("Error reading response body: %v", err)
-	}
-	body := string(bytes)
-
-	if resp.StatusCode != http.StatusOK {
-		return "", fmt.Errorf("GET request to [%s] returned status [%d]\n%s", url, resp.StatusCode, body)
-	}
-
-	return body, nil
-}
-
-// GetURLForService returns the external URL for a service in a namespace.
-func (h *TestHelper) GetURLForService(namespace string, serviceName string) (string, error) {
-	var url string
-	err := h.RetryFor(3*time.Minute, func() error {
-		// first try fetching the url from kubectl
-		cmd := exec.Command("kubectl", "-n", namespace, "get", "svc", serviceName, "-o",
-			"jsonpath={.status.loadBalancer.ingress[0].*}:{.spec.ports[0].port}")
-		out, err := cmd.Output()
+	var body string
+	err := h.RetryFor(10*time.Second, func() error {
+		resp, err := h.httpClient.Get(url)
 		if err != nil {
-			return fmt.Errorf("kubectl get svc error: %s\n%s", out, err)
-		}
-		addr := strings.TrimSpace(string(out))
-		if !strings.HasPrefix(addr, ":") {
-			url = "http://" + addr
-			return nil
+			return err
 		}
 
-		// fallback to minikube
-		cmd = exec.Command("minikube", "-n", namespace, "service", serviceName, "--url")
-		out, err = cmd.Output()
+		defer resp.Body.Close()
+		bytes, err := ioutil.ReadAll(resp.Body)
 		if err != nil {
-			return fmt.Errorf("minikube service error: %s\n%s", out, err)
+			return fmt.Errorf("Error reading response body: %v", err)
 		}
-		url = strings.TrimSpace(string(out))
+		body = string(bytes)
+
+		if resp.StatusCode != http.StatusOK {
+			return fmt.Errorf("GET request to [%s] returned status [%d]\n%s", url, resp.StatusCode, body)
+		}
+
 		return nil
 	})
-	return url, err
+
+	return body, err
 }

